clippy_lints/src/macros_hiding_unsafe_code.rs

```diff
@@ -0,0 +1,140 @@
+use clippy_utils::{diagnostics::span_lint_and_sugg, in_macro, is_lint_allowed, source::snippet_with_applicability};
+use rustc_errors::Applicability;
+use rustc_hir::{intravisit::FnKind, Block, BlockCheckMode, Body, FnDecl, FnSig, HirId, UnsafeSource, Unsafety};
+use rustc_lint::{
+    builtin::{UNSAFE_OP_IN_UNSAFE_FN, UNUSED_UNSAFE},
+    LateContext, LateLintPass, LintContext,
+};
+use rustc_middle::lint::in_external_macro;
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::Span;
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for macro calls inside an `unsafe` function which expansion
+    /// contains an `unsafe` block, while the macro call is not wrapped in an `unsafe` block
+    /// itself. This lint only triggers in functions where the `unsafe_op_in_unsafe_fn` lint is
+    /// enabled.
+    ///
+    /// **Why is this bad?** This hides an unsafe operation inside a macro call. This is against
+    /// the intention of the `unsafe_op_in_unsafe_fn` lint, which is meant to make unsafe code more
+    /// visible by requiring `unsafe` blocks also in `unsafe` functions.
+    ///
+    /// **Known problems:**
+    /// - In some cases the intention of the macro is to actually hide the unsafety, because the
+    /// macro itself ensures the safety of the `unsafe` operation.
+    /// - For local macros, either an `#[allow(unused_unsafe)]` has to be added to the new unsafe
```

Comment:
IMHO, suggesting introducing an unused unsafe block is not a good thing to do; and suggesting `allow(unused_unsafe)` is worse.

Imagine the following chain of actions:

1. Someone introduces a macro with an unsafe code.
2. In the use places it gets wrapped with `#[allow(unused_unsafe)] unsafe { ... }`.
3. Unsafe code is removed from macro.

There is a good  chance that this `unused_unsafe` will remain there unnoticed.

Reply:
Yes, I also think this isn't nice. But I don't really see how else to handle this, without rendering the lint completely useless.

Reply:
When implementing the lint and noticing, that this is necessary, I also thought about whether we should allow lints that contradict `rustc` lints at all. We have lints contradicting other Clippy lints in the `restriction` group. But we don't have lints the contradict `rustc` lints (yet).